File: 백준/Gold/17307. 색깔 통일하기/색깔 통일하기.py
"""
관찰
모든 버튼은 누르다보면 색깔이 바뀐다
그럼 몇번만에?

6 4
1 2 3 3 2 1
에서 3번 버튼을 3번눌러야 2와 같은 색이 된다 (3+x)%4 = 2 .. x = 3

그럼 N번 순회하면서 색깔이 바뀌는 지점에 몇번눌러야하는지 체크가능

누적으로 몇번 바꿔야 끝까지 가는지 더해주기 (누적합?)
끝까지 합치는 경우의 수 = max(왼쪽 끝까지 합는 경우의 수, 오른쪽 끝까지 합치는 경우의 수)
"""
import sys

input = lambda: sys.stdin.readline().strip()

N, M = map(int, input().split())
lst = list(map(int, input().split()))

delta = [[0, 0] for _ in range(N)]
cnt = [[0, 0] for _ in range(N)]

for i in range(N-1):
    delta[i][1] = (lst[i+1] - lst[i] + M) % M
    delta[i+1][0] = (lst[i] - lst[i+1] + M) % M

for i in range(1, N):
    cnt[i][0] = cnt[i-1][0] + delta[i][0]

for i in range(N-1)[::-1]:
    cnt[i][1] = cnt[i+1][1] + delta[i][1]
answer = max(cnt[0][0], cnt[0][1])
index = 1
i = 0
for a, b in cnt:
    if max(a, b) < answer:
        answer = max(a, b)
        index = i + 1
    i += 1

print(index)
print(answer)

File: 백준/Gold/17307. 색깔 통일하기/색깔 통일하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();
    private static int N, M;
    private static int[] lst;
    private static int[][] delta;
    private static long[][] cnt;

    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine());
        N  = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        lst = new int[N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            lst[i] = Integer.parseInt(st.nextToken());
        } // 입력 완료

        delta = new int[N][2];

        for (int i = 0; i < N-1; i++) {
            delta[i+1][0] = (lst[i] - lst[i+1] + M) % M;
            delta[i][1] = (lst[i+1] - lst[i] + M) % M;
        }

        cnt = new long[N][2];

        for (int i = 1; i < N; i++) {
            cnt[i][0] = cnt[i-1][0] + delta[i][0];
        }

        for (int i = N-2; i >= 0; i--) {
            cnt[i][1] = cnt[i+1][1] + delta[i][1];
        }

        long answer = Math.max(cnt[0][0], cnt[0][1]);
        int index = 1;
        for (int i = 0; i < N; i++) {
            if (answer > Math.max(cnt[i][0], cnt[i][1])) {
                answer = Math.max(cnt[i][0], cnt[i][1]);
                index = i + 1;
            }
        }

        System.out.println(index);
        System.out.println(answer);

    } // end of main

} // end of class